[READY] Add client logfile

micbou opened the pull request:
We have a lot of issue reports with Python exceptions that interrupt user workflow or even worse make the editor unusable and force users to restart it (e.g. issue https://github.com/Valloric/YouCompleteMe/issues/2192). We really want to avoid that but at the same time we can't just silence these exceptions because they are useful to debug the issue. Logging them to Vim `:messages` is not practical because we can't write to it without displaying messages in the status line which, in addition to distract users, may lead to various issues like the infamous `Press ENTER or type command to continue` message. This is why a logfile is needed.

For now, only server crashes are logged but more logging will be added: connection issues with the server (`ConnectTimeout`, `ReadTimeout`, etc. exceptions), UltiSnips unavailability, requests sent to the server, etc.

The behavior of the `:YcmToggleLogs` command is changed to accept multiple arguments where each argument is a logfile name. Each of these files is opened in a separate window or closed if already open. When no argument is given, the list of available logfiles is displayed to the user. Example:
```
Available logfiles are:
ycm_pz83u7.log
ycmd_23830_stderr_gf6j3i.log
ycmd_23830_stdout_gmpa_k.log
```

With this change and PR #2342, we will add the completers logfiles to the list of files that can be opened with the `:YcmToggleLogs` command. 

A bunch of tests are added that cover almost all changes introduced by this PR.

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/2430)
<!-- Reviewable:end -->

vheon commented:
Reviewed 9 of 11 files at r1, 1 of 1 files at r2.
Review status: 10 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[plugin/youcompleteme.vim, line 87 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvPVW-JhtnUFDYkYMT:-KVvPVW-JhtnUFDYkYMU:b-5vhyst) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/plugin/youcompleteme.vim#L87)):*
> ```VimL
>       \ get( g:, 'ycm_server_log_level', 'info' ) )
> 
> let g:ycm_server_keep_logfiles =
> ```

you forgot the `_server` section here

---

*[python/ycm/youcompleteme.py, line 194 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvUy1V-ER20nvAV5JT:-KVvUy1WmYFPBOxQtbEJ:by1gx2e) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/python/ycm/youcompleteme.py#L194)):*
> ```Python
>     handler = logging.FileHandler( self._client_logfile )
> 
>     # On Windows and Python prior to 3.4, file handles are inherited by child
> ```

I'm not sure on one thing: does this happen only on windows with python prior to 3.4? Or does this happen to *nix too? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 11 of 11 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[plugin/youcompleteme.vim, line 87 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvPVW-JhtnUFDYkYMT:-KVvY7Mn3vhEmch32HJU:b6ba99k) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/plugin/youcompleteme.vim#L87)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> you forgot the `_server` section here
</details>
No, it's on purpose: [the `server_keep_logfiles` option is defined by the server](https://github.com/Valloric/ycmd/blob/master/ycmd/default_settings.json#L41) so it would require a change in ycmd to be able to remove the `server_` part here.

---

*[python/ycm/youcompleteme.py, line 194 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvUy1V-ER20nvAV5JT:-KVvYKVH1IYwIbnJ2S2w:brbu3e7) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/python/ycm/youcompleteme.py#L194)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> I'm not sure on one thing: does this happen only on windows with python prior to 3.4? Or does this happen to *nix too? 
</details>
Only on Windows. Should I replace the `and` by `with`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[plugin/youcompleteme.vim, line 87 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvPVW-JhtnUFDYkYMT:-KVveqJyOgPaeFc4drkF:bto8gbx) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/plugin/youcompleteme.vim#L87)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> No, it's on purpose: [the `server_keep_logfiles` option is defined by the server](https://github.com/Valloric/ycmd/blob/master/ycmd/default_settings.json#L41) so it would require a change in ycmd to be able to remove the `server_` part here.
</details>
In the documentation you change it though :confused: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430)*
<!-- Sent from Reviewable.io -->

micbou commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[plugin/youcompleteme.vim, line 87 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvPVW-JhtnUFDYkYMT:-KVvfqMcwuqBvufIChHT:byeojes) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/plugin/youcompleteme.vim#L87)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> In the documentation you change it though :confused: 
</details>
Sorry I wasn't clear. We still need to define the `g:ycm_server_keep_logfiles` option because it is used by ycmd under the `server_keep_logfiles` name. However, renaming this option to `g:ycm_keep_logfiles` makes more sense when it also affects YCM logfile. So, we set the `g:ycm_server_keep_logfiles` option to the user value of `g:ycm_keep_logfiles` and, if this variable does not exist, we fall back to the value of `g:ycm_server_keep_logfiles` for backward compatibility. This effectively makes `g:ycm_keep_logfiles` the new name of this option (because it takes precedence over `g:ycm_server_keep_logfiles`).

I guess renaming [the `server_keep_logfiles` option](https://github.com/Valloric/ycmd/blob/master/ycmd/default_settings.json#L41) in ycmd to `keep_logfiles` would make this less confusing. I can prepare a PR for that and then I'll make the appropriate changes here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[plugin/youcompleteme.vim, line 87 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvPVW-JhtnUFDYkYMT:-KVw3te9vkb999k-4ZQ0:bitt80y) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/plugin/youcompleteme.vim#L87)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Sorry I wasn't clear. We still need to define the `g:ycm_server_keep_logfiles` option because it is used by ycmd under the `server_keep_logfiles` name. However, renaming this option to `g:ycm_keep_logfiles` makes more sense when it also affects YCM logfile. So, we set the `g:ycm_server_keep_logfiles` option to the user value of `g:ycm_keep_logfiles` and, if this variable does not exist, we fall back to the value of `g:ycm_server_keep_logfiles` for backward compatibility. This effectively makes `g:ycm_keep_logfiles` the new name of this option (because it takes precedence over `g:ycm_server_keep_logfiles`).
> 
> I guess renaming [the `server_keep_logfiles` option](https://github.com/Valloric/ycmd/blob/master/ycmd/default_settings.json#L41) in ycmd to `keep_logfiles` would make this less confusing. I can prepare a PR for that and then I'll make the appropriate changes here.
</details>
Oh, ok now I get it, thanks! :+1: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/youcompleteme.py, line 194 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvUy1V-ER20nvAV5JT:-KVw4736uQy61HjW1c_U:b-4s7gwn) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/python/ycm/youcompleteme.py#L194)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Only on Windows. Should I replace the `and` by `with`?
</details>
I'm not a native English speaker so I left this to the others :stuck_out_tongue_winking_eye: 

---

*[python/ycm/tests/youcompleteme_test.py, line 161 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVw59ea-kJXlC5ERx_6:-KVw59ea-kJXlC5ERx_7:b-15rcgl) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/python/ycm/tests/youcompleteme_test.py#L161)):*
> ```Python
> 
> @YouCompleteMeInstance()
> def YouCompleteMe_OnVimLeave_ClientLogfileIsRemoved_test( ycm ):
> ```

Should we also test the "keep logs file" scenario?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[python/ycm/tests/youcompleteme_test.py, line 161 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVw59ea-kJXlC5ERx_6:-KVzEoJyVgqtiSGJNdRU:b-r7wwg1) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/python/ycm/tests/youcompleteme_test.py#L161)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> Should we also test the "keep logs file" scenario?
</details>
Thanks to your suggestion, I found out that `assert_that( is_not( ... ) )` does not work as I expected: the assertion is always passing. I've fixed that and added a test where the client logfile is kept with the `server_keep_logfiles` option.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 1 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Agreed we should fix this. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> and thanks for the PR!

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[plugin/youcompleteme.vim, line 87 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvPVW-JhtnUFDYkYMT:-KW-2-8UVJrLBXUyOHT_:b-yv1rt9) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/plugin/youcompleteme.vim#L87)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary>
> Oh, ok now I get it, thanks! :+1: 
</details>
@micbou The way the code looks like now makes it appear as if `ycm_keep_logfiles` is the old name that is being kept around for backwards compatibility. We either need to fix this in the server or at least add a comment here explaining the potential confusion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-:-KW-2mc4geuuzGorODIu:b-htzawq)*
<!-- Sent from Reviewable.io -->

micbou commented:
This is ready again.

---

Reviewed 6 of 6 files at r5, 1 of 1 files at r6, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[plugin/youcompleteme.vim, line 87 at r3](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KVvPVW-JhtnUFDYkYMT:-KWckBXSX4uM809T9RYR:b9qzh2u) ([raw file](https://github.com/valloric/youcompleteme/blob/e68cedbf1c11b809781625e6703741dd59eb99e1/plugin/youcompleteme.vim#L87)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary>
> @micbou The way the code looks like now makes it appear as if `ycm_keep_logfiles` is the old name that is being kept around for backwards compatibility. We either need to fix this in the server or at least add a comment here explaining the potential confusion.
</details>
We now use the name `ycm_keep_logfiles` and set the `server_keep_logfiles` option before starting the ycmd server. Should be less confusing.

---

*[python/ycm/vimsupport.py, line 1000 at r6](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KWjByb25mFqhmF7Bqks:-KWjByb34qkjrubZTIeJ:b-ryfubq) ([raw file](https://github.com/valloric/youcompleteme/blob/4f2494e87ebd5a6e9b5dc10a436d4d943f137fe6/python/ycm/vimsupport.py#L1000)):*
> ```Python
> 
>   # Open the file
>   CheckFilename( filename )
> ```

This check is unnecessary. If the file cannot be opened then we should let Vim handle that.

---

*[python/ycm/vimsupport.py, line 951 at r7](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-KWjFbCxizwn9SLOlRek:-KWjFbCxizwn9SLOlRel:btb7j9c) ([raw file](https://github.com/valloric/youcompleteme/blob/fbe53de5a2fc7ca41602dbf7d23743f6bec8729f/python/ycm/vimsupport.py#L951)):*
> ```Python
>     raise RuntimeError(
>       "filename '{0}' cannot be opened. {1}.".format( filename,
>                                                       error.strerror ) )
> ```

This function is not needed anymore.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-:-KWjNJXRJtnf0Ye_2wiC:b-upplto)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Still <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> from me.

---

Review status: 10 of 12 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-:-KX8aiT761FJbhZ-BpN-:bgctrqv)*
<!-- Sent from Reviewable.io -->

vheon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> @homu r=Valloric

---

Reviewed 1 of 11 files at r1, 1 of 6 files at r5, 4 of 5 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/2430#-:-KXaP2VDyr1z2pakmYo5:b9nem37)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit b93c1fd has been approved by `Valloric`

<!-- @homu r=Valloric b93c1fd47c23ce57b72ba910b5d0a1ed60a057e0 -->

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/YouCompleteMe/builds/179216748)
